### PR TITLE
envoy: optimise getWildcardNetworkPolicyRule()

### PR DESCRIPTION
**Pull request by jrajahalme**
Introduce `identity.NumericIdentitySlice` to be used instead of bare `[]identity.NumericIdentity`. This allows use of the new `NumericIdentitySlice.AsUint32Slice()` function, that returns the slice of numeric identities as a slice of `uint32`s without allocating and copying memory. This is possible since `identity.NumericIdentity` is defined as `uint32`.

Use this new `NumericIdentitySlice.AsUint32Slice()` in Envoy policy updates to make them more efficient.

Optimise `getWildcardNetworkPolicyRule()` by not using a map to dedup numeric identities, but using the new generic `SortAndDedupSlice()` to sort and deduplicate in place.

### CODE REVIEW

File: pkg/identity/numericidentity.go [hunk 1]
```diff
@@ -482,6 +483,14 @@ type NumericIdentity uint32
 // MaxNumericIdentity is the maximum value of a NumericIdentity.
 const MaxNumericIdentity = math.MaxUint32
 
+type NumericIdentitySlice []NumericIdentity
+
+// AsUint32Slice returns the NumericIdentitySlice as a slice of uint32 without copying any data.
+// This is safe as long as the underlying type stays as uint32.
+func (nids NumericIdentitySlice) AsUint32Slice() []uint32 {
+	return unsafe.Slice((*uint32)(unsafe.Pointer(&nids[0])), len(nids))
```
Comment (nebril) [inline pkg/identity/numericidentity.go]: This will panic if the slice is empty. Should we allocate new slice in that case?

**Review by nebril — CHANGES_REQUESTED** [review]
LGTM overall, one small issue found inline.

File: pkg/identity/numericidentity.go (hunk 1 above)
Comment (jrajahalme) [reply pkg/identity/numericidentity.go]: I'll add a check and return a nil slice in if the slice is empty or nil!

Comment (jrajahalme) [reply pkg/identity/numericidentity.go]: Current use never called this with an empty slice. I adjusted use in `pkg/envoy/xds_server.go` a bit to make this clearer, too.

**Comment by jrajahalme** [comment]
Addressed comments, rebased!

**Comment by jrajahalme** [comment]
closed accidentally, reopening...

File: pkg/identity/numericidentity.go [hunk 2]
```diff
@@ -482,6 +483,17 @@ type NumericIdentity uint32
...
 
+type NumericIdentitySlice []NumericIdentity
+
+// AsUint32Slice returns the NumericIdentitySlice as a slice of uint32 without copying any data.
+// This is safe as long as the underlying type stays as uint32.
+func (nids NumericIdentitySlice) AsUint32Slice() []uint32 {
+	if len(nids) == 0 {
+		return nil
+	}
+	return unsafe.Slice((*uint32)(unsafe.Pointer(&nids[0])), len(nids))
```
Comment (mhofstetter) [inline pkg/identity/numericidentity.go]: conversion to `unsafe.Pointer` can be removed?

-> `return unsafe.Slice((*uint32)(&nids[0]), len(nids))`

File: pkg/identity/numericidentity.go [hunk 3]
```diff
@@ -482,6 +483,17 @@ type NumericIdentity uint32
 // MaxNumericIdentity is the maximum value of a NumericIdentity.
 const MaxNumericIdentity = math.MaxUint32
 
+type NumericIdentitySlice []NumericIdentity
+
+// AsUint32Slice returns the NumericIdentitySlice as a slice of uint32 without copying any data.
+// This is safe as long as the underlying type stays as uint32.
+func (nids NumericIdentitySlice) AsUint32Slice() []uint32 {
```
Comment (mhofstetter) [inline pkg/identity/numericidentity.go:489]: maybe add a unit-test

File: pkg/envoy/xds_server.go [hunk 4]
```diff
@@ -1384,6 +1385,28 @@ func getPortNetworkPolicyRule(sel policy.CachedSelector, wildcard bool, l7Parser
...
+func SortAndDedupSlice[K cmp.Ordered](slice []K) []K {
+	if len(slice) < 2 {
+		return slice
+	}
+
+	sort.Slice(slice, func(i, j int) bool {
+		return slice[i] < slice[j]
+	})
+
+	var n int = 1
```
Comment (mhofstetter) [inline pkg/envoy/xds_server.go]: type can be omitted

`var n = 1` or `n := 1`

File: pkg/envoy/xds_server.go [hunk 5]
```diff
@@ -1384,6 +1385,28 @@ func getPortNetworkPolicyRule(sel policy.CachedSelector, wildcard bool, l7Parser
...
 }
 
+// SortAndDedupSlice sorts the given slice and then removes duplicates from it without allocating
+// memory to do so.
+func SortAndDedupSlice[K cmp.Ordered](slice []K) []K {
+	if len(slice) < 2 {
+		return slice
+	}
+
+	sort.Slice(slice, func(i, j int) bool {
```
Comment (mhofstetter) [inline pkg/envoy/xds_server.go]: use `slices.Sort(slice)` from the new slices package

File: pkg/envoy/xds_server.go [hunk 6]
```diff
@@ -1384,6 +1385,28 @@ func getPortNetworkPolicyRule(sel policy.CachedSelector, wildcard bool, l7Parser
 	return r, canShortCircuit
 }
 
+// SortAndDedupSlice sorts the given slice and then removes duplicates from it without allocating
+// memory to do so.
+func SortAndDedupSlice[K cmp.Ordered](slice []K) []K {
```
Comment (mhofstetter) [inline pkg/envoy/xds_server.go]: i guess such a generic method either belongs to a different package than `envoy` or doesn't need to be exposed.

or try to get rid of it at all by directly use the suggested `slices.xxx` functions inline?

File: pkg/envoy/xds_server.go [hunk 7]
```diff
@@ -1384,6 +1385,28 @@ func getPortNetworkPolicyRule(sel policy.CachedSelector, wildcard bool, l7Parser
...
+	if len(slice) < 2 {
+		return slice
+	}
+
+	sort.Slice(slice, func(i, j int) bool {
+		return slice[i] < slice[j]
+	})
+
+	var n int = 1
+	for i := 1; i < len(slice); i++ {
```
Comment (mhofstetter) [inline pkg/envoy/xds_server.go]: use `slices.Compact(slice)` from the new slices package to make a unique slice

-> `return slices.Compact(slice)`

File: pkg/envoy/xds_server.go [hunk 8]
```diff
@@ -1384,6 +1385,28 @@ func getPortNetworkPolicyRule(sel policy.CachedSelector, wildcard bool, l7Parser
 	return r, canShortCircuit
 }
 
+// SortAndDedupSlice sorts the given slice and then removes duplicates from it without allocating
+// memory to do so.
+func SortAndDedupSlice[K cmp.Ordered](slice []K) []K {
+	if len(slice) < 2 {
```
Comment (mhofstetter) [inline pkg/envoy/xds_server.go]: by using `slices.xxx` functions, this additional check shouldn't be necessary.

**Review by mhofstetter — CHANGES_REQUESTED** [review]
change by itself looks good. but added suggestions to replace some of the slice-handling stuff with new Go built-ins.

File: pkg/envoy/xds_server.go [hunk 9]
```diff
@@ -1398,58 +1421,55 @@ func getWildcardNetworkPolicyRule(selectors policy.L7DataMap) *cilium.PortNetwor
...
+	} else if count == 0 {
 		// No remote policies would match this rule. Discard it.
 		return nil
+	} else {
+		// allocate slice and copy selected identities
+		remotePolicies = make([]uint32, count)
+		var i int
+		for _, selections := range sels {
+			i += copy(remotePolicies[i:], selections)
+		}
```
Comment (joamaki) [inline pkg/envoy/xds_server.go]:
```suggestion
		remotePolicies = make([]uint32, 0, count)
		for _, selections := range sels {
			remotePolicies = append(remotePolicies, selections...)
		}
```

https://go.dev/doc/effective_go#append

File: pkg/identity/numericidentity.go (hunk 2 above)
Comment (jrajahalme) [reply pkg/identity/numericidentity.go]: Thanks, makes this easier to understand and less prone for regressions!

File: pkg/envoy/xds_server.go (hunk 6 above)
Comment (jrajahalme) [reply pkg/envoy/xds_server.go]: Will get rid of this, did not know about the new slices package!

File: pkg/identity/numericidentity.go (hunk 3 above)
Comment (jrajahalme) [reply pkg/identity/numericidentity.go:489]: Added!

**Review by nebril — APPROVED** [review]

File: pkg/identity/numericidentity_test.go [hunk 10]
```diff
@@ -69,3 +70,17 @@ func TestGetAllReservedIdentities(t *testing.T) {
...
+func TestAsUint32Slice(t *testing.T) {
+	nids := NumericIdentitySlice{2, 42, 42, 1, 1024, 1}
+	uint32Slice := nids.AsUint32Slice()
+	require.NotNil(t, uint32Slice)
+	require.Len(t, uint32Slice, len(nids))
+	for i, nid := range nids {
+		require.Equal(t, nid.Uint32(), uint32Slice[i])
+	}
+
+	slices.Sort(uint32Slice)
```
Comment (mhofstetter) [inline pkg/identity/numericidentity_test.go]: nit: the last lines are actually testing Go's slicing package. Might not really the right place to test this here.

What you actually want to do is adding an additional test for `xdsServer.getPortNetworkPolicyRule` that is using this code

**Review by mhofstetter — APPROVED** [review]

**Review by aditighag — COMMENTED** [review]
Cool! Could you confirm if it's safe for `NumericIdentitySlice` to reference the same memory as the passed `[]NumericIdentity`? Earlier, we used to copy the results, so this didn't matter. 
It's not obvious to me why this change is safe without any additional locking. 
There is a note that says the slice must not be modified - https://github.com/cilium/cilium/blob/main/pkg/policy/selectorcache.go#L34-L34. How do we ensure this contract holds true - https://go.dev/play/p/j1lRJQ4jSP2?

**Review by joamaki — APPROVED** [review]

File: pkg/envoy/xds_server.go (hunk 9 above)
Comment (jrajahalme) [reply pkg/envoy/xds_server.go]: Thanks for this! I had assumed copy() be faster, but it turns out it is not! Changed as suggested.

File: pkg/identity/numericidentity_test.go (hunk 10 above)
Comment (jrajahalme) [reply pkg/identity/numericidentity_test.go]: Removed the lines using the slices package.

Comment (jrajahalme) [reply pkg/identity/numericidentity_test.go]: Added a new test for `getWildcardNetworkPolicyRule()`

**Comment by jrajahalme** [comment]
> Cool! Could you confirm if it's safe for `NumericIdentitySlice` to reference the same memory as the passed `[]NumericIdentity`? Earlier, we used to copy the results, so this didn't matter. It's not obvious to me why this change is safe without any additional locking. There is a note that says the slice must not be modified - https://github.com/cilium/cilium/blob/main/pkg/policy/selectorcache.go#L34-L34. How do we ensure this contract holds true - https://go.dev/play/p/j1lRJQ4jSP2?

Slice returned by `GetSelections()` is promised (by the selector cache) to not be modified, so no locking is needed. By convention none of the recipients should then modify the returned slice either. Unfortunately Go does not give us any tools in enforcing this invariant. 

In this particular case, we add the use the underlying slice in Network Policy resources for Envoy in our xDS cache. These resources are by design not to be modified once placed in the cache, so we should be fine here.

**Review by aditighag — APPROVED** [review]
(I was added as a reviewer as part of the sig-policy group.)

**Review by youngnick — APPROVED** [review]

**Review by rolinh — APPROVED** [review]
Not sure why I'm being pulled in for vendor changes but given there are non, lgtm in that respect :sweat_smile:

**Comment by jrajahalme** [comment]
Rebased to resolve merge conflict due to #27805 merged in the meanwhile.

**Pull request merged**